core/src/main/java/aima/core/search/api/SearchForActionsBidirectionallyFunction.java

```diff
@@ -1,12 +1,14 @@
 package aima.core.search.api;
 
-import java.util.function.BiFunction;
+import aima.core.util.datastructure.Pair;
+
+import java.util.List;
+import java.util.function.Function;
 
 /**
  * @author manthan
  */
 @FunctionalInterface
-public interface SearchForActionsBidirectionallyFunction<A, S> extends BiFunction<Problem<A, S>, Problem<A, S>, BidirectionalActions<A>> {
-    @Override
-    BidirectionalActions<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem);
+public interface SearchForActionsBidirectionallyFunction<A, S> extends Function<Pair<Problem<A,
+    S>, Problem<A, S>>, List<A>> {
```

Comment:
I am not sure of the advantage of switching from extending BiFunction to Function. I realize it works in the test case scenario in that the factory returns a Pair and therefore its just simple to pass that to the function. However, in general and in this case I don't think it makes sense to squash the arguments into a Pair structure.

As regards your concerns about BidirectionalActions. I take your point that what you want, in most cases, is just the solution - i.e. the list of actions from the start to the goal. However, this assumes traversal always from 1 direction. This may not always be the case. For example, what if you want to develop an route finding application that gives directions to two people to meet in the middle. Or more simplify if you want to build a GUI demonstrating how the algorithm works. Therefore, for now anyway, I prefer that this API stay as is.